debezium-core/src/main/java/io/debezium/transforms/ExtractNewRecordState.java

```diff
@@ -35,23 +43,29 @@
  * <p>
  * The SMT by default drops the tombstone message created by Debezium and converts the delete message into
  * a tombstone message that can be dropped, too, if required.
- *
+ * * <p>
+ * The SMT also has the option to insert fields from the original record's 'source' struct into the new 
+ * unwrapped record prefixed with "__" (for example __lsn in Postgres, or __file in MySQL)
+ * 
  * @param <R> the subtype of {@link ConnectRecord} on which this transformation will operate
  * @author Jiri Pechanec
  */
 public class ExtractNewRecordState<R extends ConnectRecord<R>> implements Transformation<R> {
 
     private static final String ENVELOPE_SCHEMA_NAME_SUFFIX = ".Envelope";
+    private static final String PURPOSE = "source field insertion";
 
     private static final Logger LOGGER = LoggerFactory.getLogger(ExtractNewRecordState.class);
 
     private boolean dropTombstones;
     private DeleteHandling handleDeletes;
     private boolean addOperationHeader;
+    private String[] addSourceFields;
     private final ExtractField<R> afterDelegate = new ExtractField.Value<R>();
     private final ExtractField<R> beforeDelegate = new ExtractField.Value<R>();
     private final InsertField<R> removedDelegate = new InsertField.Value<R>();
     private final InsertField<R> updatedDelegate = new InsertField.Value<R>();
+    private Cache<Schema, Schema> schemaUpdateCache;
```

Comment:
@adriank-convoy Could you please replace the cache with `io.debezium.util.BoundedConcurrentHashMap<K, V>` that we use in our project?